allennlp/data/dataset_readers/wikitables/type_declaration.py

```diff
@@ -0,0 +1,447 @@
+        action sequences.
+        2) Override ``resolve`` to resolve the type appropriately (see the docstring in ``resolve`` for more
+        information).
+    """
+    @property
+    def _signature(self):
+        raise NotImplementedError
+
+    @overrides
+    def resolve(self, other):
+        """
+        This method is central to type inference and checking. When a variable's type is being checked, we
+        compare what we know of its type against what is expected of its type by its context. The expectation
+        is provided as ``other``. We make sure that there are no contradictions between this type and other,
+        and return an updated type which may be more specific than the original type.
+
+        For example, say this type is of the function variable F in F(cell), and we start out with <?, d> (that
+        is, it takes any type and returns d). Now we have already resolved `cell` to be of type `e`. Then
+        ``resolve`` gets called with other = <e, ?>, because we know F is a function that took a constant of
+        type `e`. When we resolve <e, ?> against <?, d>, there will not be a contradiction, because any type
+        can be successfully resolved against ?. Finally we return <e, d> as the resolved type.
+
+        As a counter example, if we are trying to resolve <?, d> against <?, e>, the resolution fails, and in
+        that case, this method returns ``None``.
+
+        Note that resolution may be unidirectional because of ?, and so in the subclasses of this type, we
+        explicitly resolve in both directions.
+        """
+        raise NotImplementedError
+
+    @overrides
+    def __eq__(self, other):
+        return self.__class__ == other.__class__
+
+    @overrides
+    def matches(self, other):
+        return self == other or self == ANY_TYPE or other == ANY_TYPE
+
+    @overrides
+    def __str__(self):
+        if self == ANY_TYPE:
+            return "%s" % ANY_TYPE
+        else:
+            return self._signature
+
+    @overrides
+    def str(self):
+        if self == ANY_TYPE:
+            return ANY_TYPE.str()
+        else:
+            return self._signature
+
+
+class ReverseType(PlaceholderType):
+    """
+    ReverseType is a kind of ``PlaceholderType`` where type resolution involves matching the return
+    type with the reverse of the argument type. So all we care about are the types of the surrounding
+    expressions, and return a resolution that matches whatever parts are present in the type signatures
+    of the arguments and the return expressions.
+
+    Following are the resolutions for some example type signatures being matched against:
+        <?, <e,r>>      :   <<r,e>, <e,r>>
+        <<r,?>, <e,?>>  :   <<r,e>, <e,r>>
+        <<r,?>, ?>      :   <<r,?>, <?,r>>>
+        <<r,?>, <?,e>>  :   None  (causes resolution failure)
+    """
+    @property
+    def _signature(self):
+        return "<<#1,#2>,<#2,#1>>"
+
+    @overrides
+    def resolve(self, other):
```

Comment:
Since its signature is `<<#1,#2>,<#2,#1>>` no information about types in `self` is relevant. All that matters is that `other.fiirst` resolves against the reverse of `other.second` and vice versa.